### PR TITLE
Migrating from extension methods to trait methods with context bounds

### PR DESCRIPTION
**Pull request by LGLO**
This PR covers removing extension methods and moving them to trait:
* `absolve`
* `flatten`
* `when`
* `whenM`
* `unsandbox`
* `raceAll`
* `supervise`

Besides `raceAll` and `supervise` methods of `IO` companion objects are removed.

### CODE REVIEW

**Comment by NeQuissimus** [comment]
I don't really care for any of the `IO.xxx` methods if they can be replaced with `io.xxx`.

**Comment by jdegoes** [comment]
@LGLO I think I would be conservative and have the IO methods delegate to methods in the IO companion object. Why? Because at the very least, construction of the implicit evidence parameters (and associated mapping / left mapping) will impose performance overhead that's not strictly necessary, so internally in ZIO, it should always use the companion object methods. The others exist for syntactical reasons but will have overhead associated with them (which is fine for users, but not ZIO).

Does that seem like a good tradeoff?

File: core/jvm/src/test/scala/scalaz/zio/IOSpec.scala [hunk 1]
```diff
@@ -103,4 +110,89 @@ class IOSpec(implicit ee: org.specs2.concurrent.ExecutionEnv) extends AbstractRT
     unsafeRun(IO.done(terminated)) must throwA(FiberFailure(Unchecked(error)))
     unsafeRun(IO.done(failed)) must throwA(FiberFailure(Checked(error)))
   }
+
+  def testWhen = {
+    var effect: Int              = 0
```
Comment (jdegoes) [inline core/jvm/src/test/scala/scalaz/zio/IOSpec.scala]: Note these `var` can be replaced by `Ref` to keep the test more purely functional.

We're evolving to a style where there's just one `unsafeRun` per test.

File: core/jvm/src/test/scala/scalaz/zio/IOSpec.scala [hunk 2]
```diff
@@ -103,4 +110,89 @@ class IOSpec(implicit ee: org.specs2.concurrent.ExecutionEnv) extends AbstractRT
...
+    effect must_=== 1
+
+    val failure                     = new Exception("expected")
+    val failed: IO[Exception, Unit] = IO.fail(failure)
+    unsafeRun(failed.when(false)) must_=== (())
+    unsafeRun(failed.when(true)) must throwA(FiberFailure(Checked(failure)))
+  }
+
+  def testWhenM = {
+    var effect: Int              = 0
```
Comment (jdegoes) [inline core/jvm/src/test/scala/scalaz/zio/IOSpec.scala]: Ditto here.

File: core/jvm/src/test/scala/scalaz/zio/IOSpec.scala (hunk 1 above)
Comment (jdegoes) [reply core/jvm/src/test/scala/scalaz/zio/IOSpec.scala]: i.e. you can do:

```scala
unsafeRun(
....
(a must_=== b) and ...
)
```

**Comment by LGLO** [comment]
Done.

**Comment by jdegoes** [comment]
giphy

**Pull request merged**